swift/extractor/Extractor.h

```diff
@@ -0,0 +1,24 @@
+#ifndef SWIFT_EXTRACTOR_H_
+#define SWIFT_EXTRACTOR_H_
+
+#include "Configuration.h"
+#include <swift/AST/SourceFile.h>
+#include <swift/Frontend/Frontend.h>
+#include <memory>
+
+namespace codeql {
+
+class Extractor {
+ public:
+  explicit Extractor(const Configuration& config, swift::CompilerInstance& instance);
+  void extract();
+
+ private:
+  void extractFile(swift::SourceFile& file);
+
+  const Configuration& config;
```

Comment:
I'd prefer to stick to references and use "borrowing" semantics here.

Reply:
Could you elaborate on the technical reasons for this preference?

The potential lifetime issues with capturing a temporary or otherwise outliving the captured object are real risks that frequently occur, especially at public API boundaries, and especially in the absence of documentation.

If this were an internal class and the exposed API were a function as per my other comment, this would be a much smaller issue (assuming the lifetime requirement were documented) both due to the much lower "surface area" (caller and callee in the same file, usually with only one call to the constructor), and for this sort of use case due to how it'd be used: the external caller would provide the object (whose lifetime will be at least that of the function call) and we'd just forward it into the object constructed just for that call and destroyed before returning from the function.

Reply:
The config object is not owned by any specific class, but shared across (potentially many) of them.
Yes. there is a risk of the lifetime issue, but I think we can take this risk.

Reply:
for the record, as I have discussed this with Alex on zoom: I would also kind of lean towards safer approaches like copies or `shared_ptr`s, on the other hand one thing that mitigates the risk, is that we are not dealing with a long-lived daemon application, for example creating and destroying services at runtime. We are within an application that does a specific task and then exits, which makes lifetime issues somewhat simpler. In such a scenario, it's not so weird to keep in mind there is for example one single configuration object created at the beginning of `main` that is guarateed to live through the whole run (as long as we avoid objects with static storage duration, which would be a bad idea anyway).

On the other hand, I wouldn't mind switching to `shared_ptr` for peace of mind, as the overhead would probably be very negligible.